man/siteconfig.1
.TH siteconfig 1 "November 2004" "siteconfig (wzdftpd)" "wzdftpd tools"
.\" Process with
.\" groff -man -Tascii siteconfig.1 
.\"
.SH NAME
siteconfig \- online configuration of FTP server (wzdftpd)
.SH SYNOPSIS
.B siteconfig [OPTION] [-m ACTION]
.SH DESCRIPTION
The
.BI siteconfig
command can be used to show or change wzdftpd's parameters online.

If no action is specified, it reads it from the standard input.
.SH OPTIONS
.TP 12
.B \-h
Display a short usage description, including all available options.
.TP
.BI \-u " XXX"
Use username \fIXXX\fP instead of wzdftpd. 
.TP
.BI \-w " XXX"
Use password \fIXXX\fP with the username.
.TP
.BI \-h " XXX"
Connect to the specified host.
.TP
.BI \-p " XXX"
Employ the specified port number for connecting to the server.
.TP

.BI \-m " XXX"
Use action \fIXXX\fP instead of reading it from standard input.
.SH ACTION
.TP
.B site vars get [all/variable/help]
Display value of variable if the name is given, or all available variables
if parameter is all.
A contextual help is displayed if the parameter is help.
.TP
.B site vars set variable value
Change parameter to the new value.
A contextual help is displayed if the parameter is help (in this case the
second parameter is ignored).
.TP
.PD
.BR
.SH AUTHORS
.PP
wzdftpd is written and maintained by Pierre Chifflier <chifflier@cpe.fr>,
and some others contributors, full credits can be found in
.BR /usr/share/doc/wzdftpd/AUTHORS
.PD
.SH SEE ALSO
.BR wzdftpd(8), ftp(1), siteuptime(1), sitewho(1)
.PP
Full documentation on wzdftpd, including configuration and FAQs, is available at
.BR http://www.wzdftpd.net
.PP 
For help/support, try the wzdftpd mailing lists, detailed on
.BR http://sourceforge.net/mail/?group_id=78247
.PP
Report bugs at
.BR http://www.wzdftpd.net/smf/index.php


NEWS
   NEWS file for wzdftpd

0.9.0:

New features

Bugfixes

0.8.2: July 9 2007

New features

    * SQLite backend has been added
    * Brand new Windows installer
    * Extra SITE HELP documentation added for various commands
    * SQL backends can connect to SQL servers using TLS encryption 

Bugfixes

    * Lots of spelling errors and ambiguities have been corrected in server responses
    * Event triggers now work correctly on Windows when specifying full paths for the scripts
    * First entry in a directory list is now correctly sent to the client
    * MLST/MLSD command was missing the PERM= field
    * Can now set the passive port range to just one port without the server crashing
    * syslog output now works correctly
    * Users can no longer CWD into a directory they don't have list/read permissions for
    * Cookies code is now thread-safe, fixing some stability issues
    * Dozens of other bugfixes and stability improvements
    * Server will no longer randomly crash when using the "site shutdown" command

0.8.1: December 7 2006

The most important change is the migration to cmake (although it is NOT
mandatory, autoconf is still supported). Some bugs were fixed to avoid
problems with data connection on timeout, or when a file is overwritten by
a smaller one.
Added support for BNC (using IDNT), and events WIPE and PREWIPE.

0.8.0: September 7 2006

New events have been added (WIPE, PREMKDIR), and the scripts are now able to
deny a transfer. Users and groups management has been rewritten and
enhanced. SQL backends support auto-reconnection if interrupted.
Many bugs were fixed, particularly for MLSD and SITE vars commands.
Documentation has been removed from package, the documentation is now
located on the wiki: http://www.wzdftpd.net/wiki/

0.7.0: April 24 2006

Enable IPv6 and MLSD by default, add new options to reject users at connection
time. There were improvements in the Zeroconf support, events, custom commands,
unit tests, and several bugs were fixed.
Lots of deprecated or unsafe code was removed, including the support for the
previous config file format and for old backends.

0.6.0: Dec 26 2005

Major update: config file changed, multiple IPv4 and IPv6 binds,
zeroconf support, multiples fixes in GnuTLS, cookies, permissions, and server
replies. IPv6 code was also greatly improved.

0.5.5: Sep 26 2005

Security update: SITE command Remote Command Injection Vulnerability (see
wzdftpd advisory 2005-01)

0.5.4: Jul 09 2005

Better error logging, new log system based on channels. Important changes on
MySQL backend (all tables renamed), see UPGRADING file for instructions.
Problems fixed: GnuTLS errors, uid problems in MySQL backend, thread-safe
problems, crontab.

0.5.3: May 18 2005

New backend using PostgreSQL to store users. A new site command (SECTIONS)
was added to list sections, and several bugs were fixed in the MySQL backend.
UTF-8 support was also improved.

0.5.2: April 18 2005

Fixed several problems with mkdir, cwd and custom site commands. Windows
installer was also improved, and icons were added.

0.5.1: March 07 2005

Fixed problems with configuration (pasv_ip, site_cmd not working, max_ul and
max_dl not saved for groups). A memory leak has been fixed after client log
out. chown, chgrp and chmod has been added to perl and tcl modules, as well as a
log for perl. An incompatibility between OpenSSL and GnuTLS is also probably
solved. Some unit tests have also been added to the project.

0.5.0: February 21 2005

Many changes: no more shared memory ! (and associated error messages ), the
mysql backend is fully working, no more limitations on numbers of users/groups,
big efforts made on security, new authentication library to unify methods:
crypt, md5, pam, SSL certificates, etc, new library to send commands from other
programs, perl backend in development, new memory model with hash tables to get
faster access to data and improve speed, many new commands: site vars,
vars_user, vars_group, killpath, etc, ,GnuTLS support, new commands: SSCN, CPSV,
HELP (secure FXP), new events: file delection, and many, many bugfixes.

0.4.4: December 07 2004

Fixed problems with symbolic links, permissions (delete) and LIST command.
Some new variables for server for passive range and speed limits; changing max
speed for users/groups/server is now applied immediatly. The 'perms' field is
now effective for users, allowing a finer control on authorized commands.

0.4.3: October 28 2004

Things are getting better, but several problems were identified, some of them
could crash the server of corrupt data.

0.4.2: October 05 2004

Again, bugfixes. Problems were detected and fixed on permissions, gadmins,
groups, and paths on windows.

0.4.1: September 15 2004

Bugfixes ! The previous rewrite of users management has broken many things,
so this release should make the server usable again.


0.4.0: August 26 2004

 Many features / parts rewritten:

* Permissions are now displayed correctly on LIST
* users and groups management has been rewritten
* uid/gid are now used, so must be set correctly
* big speed enhancement
* IP addresses does no more need prefixes (+ and -)
* TCL module updated
* Many new cookies
* Improved events

 Major features:

* new commands: STAT
* Perl module

 Work in Progress (not fully working yet):

* UTF-8 support
* MySQL backend
* config file management

 Bugfixes:

Too many to be listed here ! (see ChangeLog)


0.3.2: May 01 2004:

 As usual now, some bugfixes:

* log file empty on win32 (except if forcing use_syslog = 0)
* user can now see its own home

But the big part is the addition of some new features:
* tcl: command putlog
* new command STAT to provide a fast way to list dir
* group tagline, cookie %grouptag
* do not search UTF-8 by default in configure

And finally some new cookies to access transfered files infos:
* %lastfile{name,size,speed,time}


tools/siteconfig/siteconfig.c
#ifdef __CYGWIN__
#include <w32api/windows.h>
#else /* __CYGWIN__ */
# include <sys/types.h>
# include <sys/ipc.h>
# include <sys/shm.h>
#endif /* __CYGWIN__ */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <wzd.h>
/*#include <wzd_shm.h>*/

#define	SHM_KEY	0x1331c0d3

unsigned long key;

/* avoid bring undefined reference */
unsigned int wzd_server_uid;

wzd_config_t * config;

char *time_to_str(time_t time)
{ /* This support functionw as written by George Shearer (Dr_Delete) */

  static char workstr[100];
  unsigned short int days=(time/86400),hours,mins,secs;
  hours=((time-(days*86400))/3600);
  mins=((time-(days*86400)-(hours*3600))/60);
  secs=(time-(days*86400)-(hours*3600)-(mins*60));

  workstr[0]=(char)0;
  if(days)
    sprintf(workstr,"%dd",days);
  if(hours)
    sprintf(workstr,"%s%s%dh",workstr,(workstr[0])?", ":"",hours);
  if(mins)
    sprintf(workstr,"%s%s%dm",workstr,(workstr[0])?", ":"",mins);
  if(secs)
    sprintf(workstr,"%s%s%ds",workstr,(workstr[0])?", ":"",secs);
  if (!days && !hours && !mins && !secs)
    sprintf(workstr,"0 seconds");

  return(workstr);
}

void usage(const char *progname)
{
  fprintf(stderr,"Usage: %s [-k shm_key] get|set [paramname value]\r\n",progname);
}

int parse_args(int argc, char **argv)
{
  int opt;
  unsigned long l;
  char *ptr;

   /* please keep options ordered ! */
  while ((opt=getopt(argc, argv, "hk:")) != -1) {
    switch (opt) {
    case 'h':
      usage(argv[0]);
      return 1;
    case 'k':
      l = strtoul(optarg,&ptr,0);
      if (*ptr != '\0') {
	usage(argv[0]);
	return 1;
      }
      key = l;
      break;
    }
  }
  
  return 0;
}

void print_config(wzd_config_t * config)
{
  time_t t;

  if (config->site_closed) {
    printf("Server is CLOSED\n");
  }

  printf("Max threads allowed: %d\n",config->max_threads);

  printf("Port: %d\n",config->port);

  printf("Passive Range: %d -> %d\n",config->pasv_low_range,config->pasv_high_range);

  printf("Max dl speed: %d\n",config->global_dl_limiter.maxspeed);
  printf("Max ul speed: %d\n",config->global_ul_limiter.maxspeed);
  
  printf("Loglevel: %s\n",loglevel2str(config->loglevel));

  time(&t);
  t = t - config->server_start;
  printf("Uptime: %s\n",time_to_str(t));
}

void help_request_get(void)
{
  printf("Valid arguments are: all, loglevel, max_dl, max_threads, max_ul\n");
  printf("  pasv_low, pasv_high, port, uptime\n");
}

int request_get(const char *arg)
{
  time_t t;

  if (!arg || strlen(arg)<=0) return -1;

  if (strcasecmp(arg,"all")==0) {
    print_config(config);
    return 0;
  }
  if (strcasecmp(arg,"max_threads")==0) {
    printf("%d\n",config->max_threads);
    return 0;
  }
  if (strcasecmp(arg,"port")==0) {
    printf("%d\n",config->port);
    return 0;
  }
  if (strcasecmp(arg,"max_dl")==0) {
    printf("%d\n",config->global_dl_limiter.maxspeed);
    return 0;
  }
  if (strcasecmp(arg,"max_ul")==0) {
    printf("%d\n",config->global_ul_limiter.maxspeed);
    return 0;
  }
  if (strcasecmp(arg,"loglevel")==0) {
    printf("%s\n",loglevel2str(config->loglevel));
    return 0;
  }
  if (strcasecmp(arg,"pasv_low")==0) {
    printf("%d\n",config->pasv_low_range);
    return 0;
  }
  if (strcasecmp(arg,"pasv_high")==0) {
    printf("%d\n",config->pasv_high_range);
    return 0;
  }
  if (strcasecmp(arg,"uptime")==0) {
    time(&t);
    t = t - config->server_start;

    printf("%s\n",time_to_str(t));
    return 0;
  }

  help_request_get();

  return 1;
}

void help_request_set(void)
{
  printf("Valid arguments are:\n");
  printf("  loglevel\n");
  printf("  serverstop:  set this to 1 to stop server\n");
}

int request_set(const char *arg, const char *value)
{
  int i;

  if (!arg || strlen(arg)<=0) return -1;

  if (strcasecmp(arg,"loglevel")==0) {
    i = str2loglevel(value);
    if (i==-1) {
      printf("Invalid level\n");
      return 1;
    }
    config->loglevel = i;
    return 0;
  }
  if (strcasecmp(arg,"serverstop")==0) {
    if (strcmp(value,"1")==0) config->serverstop=1;
    return 0;
  }

  help_request_set();

  return 1;
}


int main(int argc, char *argv[])
{
  int shmid;
#ifdef __CYGWIN__
  void * handle;
  char name[256];
#endif

  /* default values */
  key = SHM_KEY;

  if (parse_args(argc,argv)) {
    usage(argv[0]);
    exit(1);
  }

#ifdef __CYGWIN__
  sprintf(name,"%lu",key-1);
  handle = OpenFileMapping(FILE_MAP_ALL_ACCESS,FALSE,name);
  if (handle == NULL)
#else
  shmid = shmget(key-1,0,0400);
  if (shmid == -1)
#endif
  {
    fprintf(stderr,"shmget failed\n");
    fprintf(stderr,"This is probably due to\n");
    fprintf(stderr,"\t* server not started\n");
    fprintf(stderr,"\t* wrong key\n");
    return -1;
  }

#ifdef __CYGWIN__
  config = MapViewOfFile(handle,FILE_MAP_ALL_ACCESS,0, 0, 0);
  if (config == NULL)
#else
  config = shmat(shmid,NULL,0);
  if (config == (void*)-1)
#endif
  {
    fprintf(stderr,"shmat failed\n");
    return -1;
  }

  /************ begin user part *************************/

  if (optind+1 >= argc) {
    usage(argv[0]);
    exit(1);
  }
  if (strcasecmp(argv[optind],"get")==0) {
    request_get(argv[optind+1]);
  }
  else if (strcasecmp(argv[optind],"set")==0) {
    if (optind+2 >= argc) {
      usage(argv[0]);
      exit(1);
    }
    request_set(argv[optind+1],argv[optind+2]);
  } else {
    usage(argv[0]);
    exit(1);
  }
/*  print_config(config);*/

  /************ end user part ***************************/

#ifdef  __CYGWIN__
  CloseHandle(handle);
#else
  shmdt(config);
#endif
  return 0;
}
